src/gui/fileactivitydialog.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (C) by Felix Weilbach <felix.weilbach@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+#include <QStringListModel>
+#include <QStyledItemDelegate>
+#include <QSpinBox>
+#include <QPainter>
+#include <QDebug>
+#include <QLoggingCategory>
+#include <QAbstractItemModel>
+#include <QPalette>
+#include <theme.h>
+
+#include "activityjob.h"
+#include "fileactivitydialog.h"
+#include "fileactivitymodel.h"
+#include "ui_fileactivitydialog.h"
+
+Q_LOGGING_CATEGORY(lcFileActivity, "nextcloud.gui.fileactivity", QtInfoMsg)
+
+namespace OCC {
+
+class FileActivityDelegate : public QStyledItemDelegate
+{
+    Q_OBJECT
+
+public:
+    FileActivityDelegate(QObject *parent = nullptr)
+        : QStyledItemDelegate(parent)
+    {
+        _iconSize *= Theme::pixelRatio();
+    }
+
+    void paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const override
+    {
+        QStyleOptionViewItem opt(option);
+        initStyleOption(&opt, index);
+
+        painter->save();
+        painter->setClipping(true);
+        painter->setClipRect(opt.rect);
+        painter->setFont(opt.font);
+
+        drawBackground(opt, painter);
+        drawBottomLine(opt, index, painter);
+        const auto contentRect = opt.rect.adjusted(_margins.left(), _margins.top(), -_margins.right(), -_margins.bottom());
+        const auto item = qvariant_cast<FileActivity>(index.data());
+        drawMessageIcon(contentRect, item, painter);
+        const auto timestampRect = drawTimestamp(opt, contentRect, item, painter);
+        drawMessage(opt, timestampRect, item, painter);
+
+        painter->restore();
+    }
+
+    QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+    {
+        QStyleOptionViewItem opt(option);
+        initStyleOption(&opt, index);
+
+        int textHeight = timestampBox(opt, index).height()
+            + _spacingVertical + messageBox(opt, index).height();
+        int iconHeight = _iconSize;
+        int h = textHeight > iconHeight ? textHeight : iconHeight;
```

Comment:
@FlexW Is this `h` necessary? What about this way:

if (textHeight < iconHeight) {
    textHeight = textHeight;
}